src/frontend/src/optimizer/rule/stream/filter_with_now_to_join_rule.rs

```diff
@@ -38,21 +37,28 @@ impl Rule for FilterWithNowToJoinRule {
 
         let mut rewriter = NowAsInputRef::new(lhs_len);
 
-        // If the `now` is not a valid dynamic filter expression,
+        // If the `now` is not a valid dynamic filter expression, we will not push it down.
         filter.predicate().conjunctions.iter().for_each(|expr| {
             if let Some((input_expr, cmp, now_expr)) = expr.as_now_comparison_cond() {
                 let now_expr = rewriter.rewrite_expr(now_expr);
-                now_filters.push(
-                    FunctionCall::new(cmp, vec![input_expr, now_expr])
-                        .unwrap()
-                        .into(),
-                );
+                now_filters.push(FunctionCall::new(cmp, vec![input_expr, now_expr]).unwrap());
             } else {
                 remainder.push(expr.clone());
             }
         });
 
-        if now_filters.is_empty() {
+        // We want to put `input_expr >/>= now_expr` before `input_expr </<= now_expr` as the former
+        // will introduce a watermark that can reduce state (since `now_expr` is monotonically
+        // increasing)
+        now_filters.sort_by_key(|l| rank_cmp(l.get_expr_type()));
+
+        // Ignore no now filter & forbid now filters that do not create a watermark
+        if now_filters.is_empty()
+            || !matches!(
+                now_filters[0].get_expr_type(),
+                Type::GreaterThan | Type::GreaterThanOrEqual
+            )
+        {
```

Comment:
we can not judge it only by the cmp expression's type, we should use WatermarkAnalyzer to analyze it monotonic.

Reply:
Watermark analyzer does not analyze cmp expr. I guess we can analyze the now_expr, even though we already ensure it is of the stricter form `now() [+- const_expr]`

Reply:
Since it is of a stricter form, I made the debug_assert that the watermark analyzer will produce a watermark on the right input column

Reply:
> we already ensure it is of the stricter form now() [+- const_expr

oh, I get it now. thanks